py/asmxtensa.c

```diff
@@ -264,4 +267,31 @@ void asm_xtensa_call_ind_win(asm_xtensa_t *as, uint idx) {
     asm_xtensa_op_callx8(as, ASM_XTENSA_REG_A8);
 }
 
+void asm_xtensa_bit_branch(asm_xtensa_t *as, mp_uint_t reg, mp_uint_t bit, mp_uint_t label, mp_uint_t condition) {
+    uint32_t dest = get_label_dest(as, label);
+    int32_t rel = dest - as->base.code_offset - 4;
+    if (as->base.pass == MP_ASM_PASS_EMIT && !SIGNED_FIT8(rel)) {
+        mp_obj_new_exception_msg_varg(&mp_type_RuntimeError, ET_OUT_OF_RANGE, MP_QSTR_bit_branch);
+    }
+    asm_xtensa_op24(as, ASM_XTENSA_ENCODE_RRI8(7, condition | ((bit >> 4) & 0x01), reg, bit & 0x0F, rel & 0xFF));
+}
+
+void asm_xtensa_call0(asm_xtensa_t *as, mp_uint_t label) {
+    uint32_t dest = get_label_dest(as, label);
+    int32_t rel = dest - as->base.code_offset - 3;
+    if (as->base.pass == MP_ASM_PASS_EMIT && (((rel & 0x3) != 0) || !SIGNED_FIT18(rel))) {
+        mp_obj_new_exception_msg_varg(&mp_type_RuntimeError, ET_OUT_OF_RANGE, MP_QSTR_call0);
+    }
+    asm_xtensa_op_call0(as, rel);
+}
+
+void asm_xtensa_l32r(asm_xtensa_t *as, mp_uint_t reg, mp_uint_t label) {
+    uint32_t dest = get_label_dest(as, label);
+    int32_t rel = dest - as->base.code_offset;
```

Comment:
Does this `rel` computation need to align things, as done in `asm_xtensa_op_l32r`?  Otherwise there could be an edge case in checking if the relative offset fits or not.

Reply:
You're right - it will be in the final iteration.  I need to add a couple more tests for `J` and `JX` opcodes, a new commit to skip emitting the prologue's initial jump skipping the constant pool if said pool is empty (as it is a possibility with inline assembler blocks) and then I guess this PR should be pretty much ready on my end.